lib/Dialect/Arc/Transforms/StripSV.cpp

```diff
@@ -0,0 +1,178 @@
+  auto mlirModule = getOperation();
+  opsToDelete.clear();
+  clockGateModuleNames.clear();
+  externModuleNames.clear();
+
+  auto expectedClockGateInputs =
+      ArrayAttr::get(&getContext(), {StringAttr::get(&getContext(), "in"),
+                                     StringAttr::get(&getContext(), "test_en"),
+                                     StringAttr::get(&getContext(), "en")});
+  auto expectedClockGateOutputs =
+      ArrayAttr::get(&getContext(), {StringAttr::get(&getContext(), "out")});
+  auto i1Type = IntegerType::get(&getContext(), 1);
+
+  for (auto extModOp : mlirModule.getOps<hw::HWModuleExternOp>()) {
+    if (extModOp.getVerilogModuleName() == "EICG_wrapper") {
+      if (extModOp.getArgNames() != expectedClockGateInputs ||
+          extModOp.getResultNames() != expectedClockGateOutputs) {
+        extModOp.emitError("clock gate module `")
+            << extModOp.moduleName() << "` has incompatible port names "
+            << extModOp.getArgNamesAttr() << " -> "
+            << extModOp.getResultNamesAttr();
+        return signalPassFailure();
+      }
+      if (extModOp.getArgumentTypes() !=
+              ArrayRef<Type>{i1Type, i1Type, i1Type} ||
+          extModOp.getResultTypes() != ArrayRef<Type>{i1Type}) {
+        extModOp.emitError("clock gate module `")
+            << extModOp.moduleName() << "` has incompatible port types "
+            << extModOp.getArgumentTypes() << " -> "
+            << extModOp.getResultTypes();
+        return signalPassFailure();
+      }
+      clockGateModuleNames.insert(extModOp.moduleNameAttr());
+    } else {
+      externModuleNames.insert(extModOp.moduleNameAttr());
+    }
+    opsToDelete.push_back(extModOp);
+  }
+  LLVM_DEBUG(llvm::dbgs() << "Found " << clockGateModuleNames.size()
+                          << " clock gates, " << externModuleNames.size()
+                          << " other extern modules\n");
+
+  // Remove `sv.*` operation attributes.
+  mlirModule.walk([](Operation *op) {
+    auto isSVAttr = [](NamedAttribute attr) {
+      return attr.getName().getValue().startswith("sv.");
+    };
+    if (llvm::any_of(op->getAttrs(), isSVAttr)) {
+      SmallVector<NamedAttribute> newAttrs;
+      newAttrs.reserve(op->getAttrs().size());
+      for (auto attr : op->getAttrs())
+        if (!isSVAttr(attr))
+          newAttrs.push_back(attr);
+      op->setAttrs(newAttrs);
+    }
+  });
+
+  // Remove ifdefs and verbatim.
+  for (auto verb : mlirModule.getOps<sv::VerbatimOp>())
+    opsToDelete.push_back(verb);
+  for (auto verb : mlirModule.getOps<sv::IfDefOp>())
+    opsToDelete.push_back(verb);
+
+  for (auto module : mlirModule.getOps<hw::HWModuleOp>()) {
+    for (Operation &op : *module.getBodyBlock()) {
+      // Remove ifdefs and verbatim.
+      if (isa<sv::IfDefOp, sv::CoverOp, sv::CoverConcurrentOp>(&op)) {
+        opsToDelete.push_back(&op);
+        continue;
+      }
+      if (isa<sv::VerbatimOp, sv::AlwaysOp>(&op)) {
```

Comment:
Ah, and here is the bold play.  It looks like you are assuming only VerbatimExpr and VerbatimWire and the side-effect form can impact simulation (and don't appear).

Reply:
This is a *very* ugly stopgap to allow us to use FIRRTL inputs for experiments with the arc dialect, until we can remove some of the SV op emission from the FIRRTL-to-HW pipeline. The idea is to hack it up here to experiment in arc, then go fix things in the other dialects such that they become easier to analyze (beyond just the needs of SV emission).